sycl/include/sycl/detail/vector_arith.hpp

```diff
@@ -0,0 +1,388 @@
+                                                                               \
+  template <typename T = DataT>                                                \
+  friend std::enable_if_t<(COND), vec_t> operator BINOP(const vec_t & Lhs,     \
+                                                      const DataT & Rhs) {     \
+    return Lhs BINOP vec_t(Rhs);                                               \
+  }                                                                            \
+  template <typename T = DataT>                                                \
+  friend std::enable_if_t<(COND), vec_t> operator BINOP(const DataT & Lhs,     \
+                                                      const vec_t & Rhs) {     \
+    return vec_t(Lhs) BINOP Rhs;                                               \
+  }                                                                            \
+  template <typename T = DataT>                                                \
+  friend std::enable_if_t<(COND), vec_t> &operator OPASSIGN(vec_t & Lhs,       \
+                                                          const vec_t & Rhs) { \
+    Lhs = Lhs BINOP Rhs;                                                       \
+    return Lhs;                                                                \
+  }                                                                            \
+  template <int Num = NumElements, typename T = DataT>                         \
+  friend std::enable_if_t<(Num != 1) && (COND), vec_t &> operator OPASSIGN(    \
+      vec_t & Lhs, const DataT & Rhs) {                                        \
+    Lhs = Lhs BINOP vec_t(Rhs);                                                \
+    return Lhs;                                                                \
+  }
+
+/****************************************************************
+ *                       vec_arith_common
+ *                 /           |             \
+ *                /            |               \
+ *     vec_arith<int>     vec_arith<float> ...   vec_arith<byte>
+ *                \            |               /
+ *                 \           |              /
+ *                        sycl::vec<T>
+ * 
+ * vec_arith_common is the base class for vec_arith. It contains
+ * the common math operators of sycl::vec for all types.
+ * vec_arith is the derived class that contains the math operators
+ * specialized for certain types. sycl::vec inherits from vec_arith.
+ * *************************************************************/
+template <typename DataT, int NumElements> class vec_arith_common;
+template <typename DataT> struct vec_helper;
+
+template <typename DataT, int NumElements>
+class vec_arith : public vec_arith_common<DataT, NumElements> {
+public:
+  using vec_t = vec<DataT, NumElements>;
+  using ocl_t = rel_t<DataT>;
+  template <typename T> using vec_data = vec_helper<T>;
```

Comment:
I worry these will leak into user-space. Should they be protected? Same goes for the byte specialization and vec_arithmetic_common.